### PR TITLE
Add server url button to main menu

### PR DESCRIPTION
**Pull request by siliconsniffer**
Add compact, short information about your PR for easier understanding:

This PR adds a button to the main menu for players to open the URL of the selected server.

- Goal of the PR
Add a button to open the URL.
- How does the PR work?
It adds a button and uses the `core.open_url_dialog` to display the URL and open it.
- Does it resolve any reported issue?
#15237
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
2.3

## How to test

Open the menu and select a server, button should show up.
Press it and see if it opens the URL.

## Preview
image
image

### CODE REVIEW

File: builtin/mainmenu/tab_online.lua [hunk 1]
```diff
@@ -174,6 +174,13 @@ local function get_formspec(tabview, name, tabdata)
 				"server_view_clients.png") .. ";btn_view_clients;]"
 		end
 
+		if selected_server.url then
+			retval = retval .. "tooltip[btn_server_url;" .. fgettext("Server URL") .. "]"
```
Comment (Zughy) [inline builtin/mainmenu/tab_online.lua:178]:
```suggestion
			retval = retval .. "tooltip[btn_server_url;" .. fgettext("Server Website") .. "]"
```

I think it's better for communication

Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:178]: I agree. Might be easier for translation too...

Comment (y5nw) [reply builtin/mainmenu/tab_online.lua]: IMO it would also be nice to display the URL in the tooltip.

Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua]: Ooops, I was too quick. Mhm, don't know the url is already shown in the dialog.

File: builtin/mainmenu/tab_online.lua [hunk 2]
```diff
@@ -362,6 +369,12 @@ local function main_button_handler(tabview, fields, name, tabdata)
 		return true
 	end
 
+	if fields.btn_server_url then
+		local url = find_selected_server().url
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:373]: could inline the `url` variable

File: builtin/mainmenu/tab_online.lua [hunk 3]
```diff
@@ -174,6 +174,13 @@ local function get_formspec(tabview, name, tabdata)
 				"server_view_clients.png") .. ";btn_view_clients;]"
 		end
 
+		if selected_server.url then
+			retval = retval .. "tooltip[btn_server_url;" .. fgettext("Server Website") .. "]"
+			retval = retval .. "style[btn_server_url;padding=6]"
+			retval = retval .. "image_button[" .. (can_view_clients_list and "4" or "4.5") .. ",1.3;0.5,0.5;" ..
+			core.formspec_escape(defaulttexturedir .. "server_url.png") .. ";btn_server_url;]"
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:181]: Misleading indentation. I believe our style guidelines would want this to be indented two times to make clear that it is a continuation of the expression on the previous line. (Though I also see some continuations use a single tab; I don't mind either way.)

**Review by appgurueu — COMMENTED** [review]
Otherwise looks good and works.

Bugs like the tooltip bug are better addressed in pure bugfix PRs.

File: builtin/mainmenu/tab_online.lua (hunk 2 above)
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:373]: Yep, I even had it like that before. Will revert :+1:

**Comment by siliconsniffer** [comment]
Sounds good. Will take a look probably in the evening. I also found another bug, I'll try to put them in one PR.

**Comment by SmallJoker** [comment]
Icon counter-proposal: (CC 0)
grafik

This is not a change request but an idea. The current icon would also need a higher resolution for better downscaling results.

**Comment by Zughy** [comment]
I find that icon too generic, I read it as it "open something in another tab".

For the scaling, OP, just take my icon and double it

**Comment by appgurueu** [comment]
To me, the globe (or "chain link") would be more indicative of, well, a weblink. The symbol Krock suggested I would associate with "open something / follow some reference" (what?). If I do some quick searches for "icons for URLs", "icons for links", the former two dominate the results for me; they seem to be more common than the one Krock suggested (though it also appears).

File: builtin/mainmenu/tab_online.lua [hunk 4]
```diff
@@ -174,12 +174,12 @@ local function get_formspec(tabview, name, tabdata)
...
-			retval = retval .. "tooltip[btn_server_url;" .. fgettext("Server Website") .. "]"
-			retval = retval .. "style[btn_server_url;padding=6]"
-			retval = retval .. "image_button[" .. (can_view_clients_list and "4" or "4.5") .. ",1.3;0.5,0.5;" ..
-			core.formspec_escape(defaulttexturedir .. "server_url.png") .. ";btn_server_url;]"
-		end
+			if selected_server.url then
+				retval = retval .. "tooltip[btn_server_url;" .. fgettext("Server Website") .. "]"
+				retval = retval .. "style[btn_server_url;padding=6]"
+				retval = retval .. "image_button[" .. (can_view_clients_list and "4" or "4.5") .. ",1.3;0.5,0.5;" ..
+				core.formspec_escape(defaulttexturedir .. "server_url.png") .. ";btn_server_url;]"
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:181]: Sorry, I think you misunderstood me. I wanted just this line (the "continuation of the expression") to be indented by one or two tabs, not the entire `if` statement.

Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:181]: Should be good now?

**Review by appgurueu — APPROVED** [review]

**Comment by siliconsniffer** [comment]
@grorp What resolution is this? Can you try it with the upscaled one again?

**Comment by siliconsniffer** [comment]
image
This is what it looks like for me when it is reduced to a quarter of my screen.

**Comment by Zughy** [comment]
Until #10733 is fixed, it'll probably still look bad no matter what. It'd be very nice if that issue were solved so that pixel art in general will actually start looking good

**Comment by siliconsniffer** [comment]
Ah, hence the upscaled didn't change anything. This PR should be ready then?

**Comment by grorp** [comment]
> Until #10733 is fixed, it'll probably still look bad no matter what. It'd be very nice if that issue were solved so that pixel art in general will actually start looking good.
> 
> (Maybe you haven't noticed, but also the profile icon is wrongly scaled: the forehead in your screen is shorter than it should be)

Actually the downscaling problem (but not #10733) was already fixed, but we had to revert the fix in #15385. With the revert reverted, it scales ok: screenshot

While the bad scaling in general is not this PR's fault, it may be possible to change the icon so that it scales better, which is why I mentioned this problem here. I dropped in the old "cdb_viewonline.png" icon for comparison, it scaled better.

**Comment by siliconsniffer** [comment]
Personally I'd like to have this solved "properly" in a separate PR instead of changing the icon. What do the others think?

File: textures/base/pack/server_url.png
Comment (sfan5) [inline textures/base/pack/server_url.png:1]: not optimized:
```
% oxipng -o max server_url.png 
Processing: server_url.png
356 bytes (64.36% smaller): server_url.png
```

File: builtin/mainmenu/tab_online.lua [hunk 5]
```diff
@@ -174,6 +174,13 @@ local function get_formspec(tabview, name, tabdata)
 				"server_view_clients.png") .. ";btn_view_clients;]"
 		end
 
+		if selected_server.url then
+			retval = retval .. "tooltip[btn_server_url;" .. fgettext("Server Website") .. "]"
```
Comment (sfan5) [inline builtin/mainmenu/tab_online.lua:178]: better like this imo
```suggestion
			retval = retval .. "tooltip[btn_server_url;" .. fgettext("Visit server website") .. "]"
```

**Review by sfan5 — CHANGES_REQUESTED** [review]

File: builtin/mainmenu/tab_online.lua (hunk 5 above)
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:178]: I disagree. Adding the "visit" makes it sound like the button would open the website directly - at least for me.

Comment (sfan5) [reply builtin/mainmenu/tab_online.lua:178]: I think button tooltips should describe the action it performs, do you have a different suggestion?

**Comment by appgurueu** [comment]
> Personally I'd like to have this solved "properly" in a separate PR instead of changing the icon. What do the others think?

I agree. I'm not yet sure how we should do that though. Should we:

- Apply bilinear downscaling to GUI images generally (but not globally, because that apparently can mess things up, see the revert)?
- Add a `style[]` option to control filtering?

**Comment by siliconsniffer** [comment]
Applying it for all GUI images would be the "cleaner" option imo.
That way it gets "forced" onto the GUI and (hopefully) creates are more consistent look.

File: builtin/mainmenu/tab_online.lua (hunk 5 above)
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:178]: Maybe "open"? Not a native speaker but for me that feels more like a term to describe the general action here.

**Review by sfan5 — APPROVED** [review]

**Pull request merged**